crates/metaslang/bindings/src/lib.rs

```diff
@@ -54,6 +55,54 @@ pub struct Bindings<KT: KindTypes + 'static> {
     context: Option<GraphHandle>,
 }
 
+pub enum FileDescriptor {
+    User(String),
+    System(String),
+}
+
+pub(crate) struct FileDescriptorError;
+
+impl FileDescriptor {
+    // Internal functions to convert a FileDescriptor to and from a string for
+    // representation inside the stack graph
+
+    pub(crate) fn as_string(&self) -> String {
+        match self {
+            Self::User(path) => format!("user:{path}"),
+            Self::System(path) => format!("system:{path}"),
+        }
+    }
+
+    pub(crate) fn from_string(value: &str) -> Result<Self, FileDescriptorError> {
+        if let Some(path) = value.strip_prefix("user:") {
+            Ok(FileDescriptor::User(path.into()))
+        } else if let Some(path) = value.strip_prefix("system:") {
+            Ok(FileDescriptor::System(path.into()))
+        } else {
+            Err(FileDescriptorError)
```

Comment:
nit: I suggest includng the `value` in the resulting `FileDescriptorError`, to appear in its `Display` for debugging.

Reply:
I'm a bit hesitant to do `unreachable!()` here since this is called from the graph building functions (`is-system-file` and `resolve-path`) and those can potentially pass any string as a parameter. If we panic here, the program will crash instead of returning a graph construction error. Granted, we're also panicking if the graph building fails, but that happens at a higher level. I'll add the value to the error.

Reply:
> If we panic here, the program will crash instead of returning a graph construction error. 

I see. Sounds good on keeping the error, but I suggested including the `value` in the error because:

1. It helps in debugging/`Display` of the error.
2. We don't need to wrap it in another "expectation"/error in callsites to show what failed, like the two callsites here.

But up to you. I don't think it is worth spending more of your time if you see otherwise.

Reply:
Right, I understand the reason. But if I try to add the `value` to the error, clippy rightly complains that it's not being used.

I think the problem here is that we need to panic when this is called from the two callsites below, but we don't when called from the graph functions. What do you think about using two functions `try_from` and `from`, like the traits? We don't want to implement the traits because then we need to expose them to the public API, but in this case I think it'd be ok to reuse those names, since they would have the expected semantics.

Reply:
Sounds good to me 👍